src/d3a/models/strategy/storage.py

```diff
@@ -226,15 +231,45 @@ def event_tick(self, *, area):
     def event_trade(self, *, market_id, trade):
         market = self.area.get_future_market_from_id(market_id)
         super().event_trade(market_id=market_id, trade=trade)
+        if trade.buyer == self.owner.name:
+            self._energy_bought_type(trade)
         if trade.offer.seller == self.owner.name:
+            self._energy_sell_type(trade)
             self.state.pledged_sell_kWh[market.time_slot] += trade.offer.energy
             self.state.offered_sell_kWh[market.time_slot] -= trade.offer.energy
 
+    def _is_local(self, trade):
+        for child in self.area.children:
+            if child.name == trade.seller:
+                return True
+
+    def _energy_sell_type(self, trade):
+        energy = trade.offer.energy
+        while limit_float_precision(energy) > 0:
+            recent_energy = self.state.get_used_storage_share[0].value
+            if energy >= recent_energy:
+                energy -= recent_energy
+                self.state.get_used_storage_share.pop(0)
+            elif energy < recent_energy:
+                last = self.state.get_used_storage_share[0]
```

Comment:
In order to avoid confusion, I suggest to omit the declaration of `last`  and put `self.state.get_used_storage_share[0]` into line 256. Does it make sense?

Reply:
Please have a look at it here: https://github.com/gridsingularity/d3a/pull/497/files#diff-6479cb83f0833c9660b171d3fd0bfc57R250